ENIGMAsystem/SHELL/Universal_System/Resources/polygon_internal.h

```diff
@@ -0,0 +1,142 @@
+/** Copyright (C) 2021 Nabeel Danish
+*** Nabeel Danish
+**/
+
+#ifndef ENIGMA_POLYGON_INTERNAL_H
+#define ENIGMA_POLYGON_INTERNAL_H
+
+#include <vector>
+#include "AssetArray.h"
+
+namespace enigma 
+{
+	#define NULL_POINT -1;
+	#define precision pow(4, -10)
+
+	// The Point class represents a point in a polygon
+	// Resource
+	class Vector2D 
+	{
+		// Attributes
+		private: 
+			double x;
+        	double y;
+
+		// Methods
+		public:
+			Vector2D();
+			Vector2D(double x, double y);
+			Vector2D(const Vector2D& obj);
+			void copy(const Vector2D& obj);
+
+			double getX();
+			double getY();
+			double getMagnitude();
+			double getAngle();
+			Vector2D getNormR();
+			Vector2D getNormL();
+			Vector2D getUnitVector();
+
+			void setX(double x);
+			void setY(double y);
+			void setMagnitude(double number);
+			void setAngle(double angle);
+			void scale(double number);
+			void invert(bool x, bool y);
+
+			Vector2D add(const Vector2D& B);
+			Vector2D minus(const Vector2D& B);
+			Vector2D rotate(double angle);
+			Vector2D interpolate(double number);
+
+			double angleBetween(Vector2D& B);
+			double dotProduct(const Vector2D& B);
+			double perpProduct(const Vector2D& B);
+			double crossProduct(const Vector2D& B);
+
+			bool equal(const Vector2D& B);
+
+			void print();
+	};
+
+	// The Polygon class represents the polygon that will be used
+	// for detecting collisions in 2D space by the collision detection 
+	// system
+	class Polygon {
+		// Attributes
+		private:
+			std::vector<Vector2D> points;
+			std::vector<std::vector<Vector2D>> subpolygons;
+			int height;
+			int width;
+			int xoffset;
+			int yoffset;
+			int numSubPolygons;
+			double scale;
+			double angle;
+
+			// Asset Array mandatory attributes
+			bool _destroyed = false;
+
+		// Methods
+		public:
+			Polygon();
+			Polygon(int size);
+			Polygon(int h, int w, int x, int y);
+			Polygon(const Vector2D* points, int size);
+			Polygon(const Polygon& obj);
+			
+			// Asset Array mandatory functions
+			static const char* getAssetTypeName() { return "Polygon"; }
+ 			bool isDestroyed() const { return _destroyed; }
+  			void destroy() { _destroyed = true; }
+
+			// Getters and Setters
+			int getNumPoints();
+			std::vector<Vector2D> getPoints();
+			std::vector<Vector2D> getPoints(double offset_x, double offset_y);
+
+			int getHeight();
+			int getWidth();
+			int getXOffset();
+			int getYOffset();
+			double getScale();
+			double getAngle();
+
+			void setHeight(int h);
+			void setWidth(int w);
+			void setXOffset(int x);
+			void setYOffset(int y);
+			void setScale(double s);
+			void setAngle(double a);
+
+			void addPoint(const Vector2D& point);
+			void addPoint(int x, int y);
+
+			void removePoint(int x, int y);
+			void removePoint(const Vector2D& point);
+
+			void copy(const Polygon& obj);
+			void copy(const Vector2D* points, int size);
+
+			std::vector<Vector2D> getNorms(double offset_x, double offset_y);
+
+			void print();
```

Comment:
This one is only here for debugging. Although I am not even using it now, so I don't think we need this method, since there isn't much sense in "printing" a polygon. I can overload the output stream operator If you want, but you would have to guide me on what to print.